File: lib/services/github_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/book.dart';

class GitHubService {
  static const String _baseUrl = 'https://api.github.com';
  static const String _tokenKey = 'github_token';
  static const String _userKey = 'github_user';

  String? _token;
  Map<String, dynamic>? _userInfo;

  // Singleton pattern
  static final GitHubService _instance = GitHubService._internal();
  factory GitHubService() => _instance;
  GitHubService._internal();

  // Check if authenticated
  bool get isAuthenticated => _token != null;

  // Load stored authentication
  Future<bool> loadStoredAuth() async {
    final prefs = await SharedPreferences.getInstance();
    _token = prefs.getString(_tokenKey);
    final userJson = prefs.getString(_userKey);
    
    if (userJson != null) {
      _userInfo = json.decode(userJson);
    }
    
    return _token != null;
  }

  // Validate and store token
  Future<Map<String, dynamic>> validateAndSetupToken(String token) async {
    if (!token.startsWith('ghp_')) {
      throw Exception('Please enter a valid GitHub personal access token');
    }

    final response = await http.get(
      Uri.parse('$_baseUrl/user'),
      headers: {
        'Authorization': 'Bearer $token',
        'Accept': 'application/vnd.github.v3+json',
        'User-Agent': 'AbsoluteScenes-Mobile',
      },
    );

    if (response.statusCode == 401) {
      throw Exception('Invalid token. Please check that you copied it correctly.');
    } else if (response.statusCode == 403) {
      throw Exception('Token lacks required permissions. Please ensure "repo" and "user:email" scopes are selected.');
    } else if (response.statusCode != 200) {
      throw Exception('GitHub API error: ${response.statusCode}');
    }

    final userInfo = json.decode(response.body);
    await _storeAuth(token, userInfo);
    return userInfo;
  }

  // Store authentication
  Future<void> _storeAuth(String token, Map<String, dynamic> userInfo) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_tokenKey, token);
    await prefs.setString(_userKey, json.encode(userInfo));
    
    _token = token;
    _userInfo = userInfo;
  }

  // Get repositories with book files
  Future<List<Repository>> getUserRepositoriesWithBooks() async {
    if (!isAuthenticated) throw Exception('Not authenticated');

    final response = await http.get(
      Uri.parse('$_baseUrl/user/repos?sort=updated&per_page=100'),
      headers: {
        'Authorization': 'Bearer $_token',
        'Accept': 'application/vnd.github.v3+json',
        'User-Agent': 'AbsoluteScenes-Mobile',
      },
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to fetch repositories: ${response.statusCode}');
    }

    final repos = json.decode(response.body) as List;
    final bookRepos = <Repository>[];

    // Check each repo for book files
    for (final repo in repos) {
      final bookFile = await _checkForBookFile(repo['full_name']);
      if (bookFile != null) {
        bookRepos.add(Repository(
          fullName: repo['full_name'],
          name: repo['name'],
          description: repo['description'],
          bookFileName: bookFile,
        ));
      }
    }

    return bookRepos;
  }

  // Check if repository contains a book file
  Future<String?> _checkForBookFile(String repoFullName) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/repos/$repoFullName/contents'),
        headers: {
          'Authorization': 'Bearer $_token',
          'Accept': 'application/vnd.github.v3+json',
          'User-Agent': 'AbsoluteScenes-Mobile',
        },
      );

      if (response.statusCode == 200) {
        final contents = json.decode(response.body) as List;
        for (final file in contents) {
          if (file['type'] == 'file' && file['name'].toString().endsWith('.book')) {
            return file['name'];
          }
        }
      }
    } catch (e) {
      // Ignore errors - repo might be empty or inaccessible
    }
    
    return null;
  }

  // Download book from repository
  Future<Book> downloadBookFromRepository(String repoFullName, String fileName) async {
    if (!isAuthenticated) throw Exception('Not authenticated');

    final response = await http.get(
      Uri.parse('$_baseUrl/repos/$repoFullName/contents/$fileName'),
      headers: {
        'Authorization': 'Bearer $_token',
        'Accept': 'application/vnd.github.v3+json',
        'User-Agent': 'AbsoluteScenes-Mobile',
      },
    );

    if (response.statusCode == 404) {
      throw Exception('Book file not found in repository');
    } else if (response.statusCode != 200) {
      throw Exception('Failed to download file: ${response.statusCode}');
    }

    final fileData = json.decode(response.body);
    final content = utf8.decode(base64.decode(fileData['content'].replaceAll(RegExp(r'\s'), '')));
    
    // Normalize line endings
    final normalizedContent = content.replaceAll('\r\n', '\n').replaceAll('\r', '\n');
    
    try {
      final bookJson = json.decode(normalizedContent);
      return Book.fromJson(bookJson);
    } catch (e) {
      throw Exception('Downloaded file is not a valid book format');
    }
  }

  // Save book to repository
  Future<void> saveBookToRepository(String repoFullName, String fileName, Book book, String commitMessage) async {
    if (!isAuthenticated) throw Exception('Not authenticated');

    // Get current file SHA if it exists
    String? fileSha;
    try {
      final fileResponse = await http.get(
        Uri.parse('$_baseUrl/repos/$repoFullName/contents/$fileName'),
        headers: {
          'Authorization': 'Bearer $_token',
          'Accept': 'application/vnd.github.v3+json',
          'User-Agent': 'AbsoluteScenes-Mobile',
        },
      );

      if (fileResponse.statusCode == 200) {
        final fileData = json.decode(fileResponse.body);
        fileSha = fileData['sha'];
      }
    } catch (e) {
      // File doesn't exist - will create new
    }

    // Prepare content
    final bookJson = json.encode(book.toJson());
    final normalizedContent = bookJson.replaceAll('\r\n', '\n').replaceAll('\r', '\n');
    final encodedContent = base64.encode(utf8.encode(normalizedContent));

    final updateData = {
      'message': commitMessage,
      'content': encodedContent,
      'branch': 'main',
    };

    if (fileSha != null) {
      updateData['sha'] = fileSha;
    }

    final response = await http.put(
      Uri.parse('$_baseUrl/repos/$repoFullName/contents/$fileName'),
      headers: {
        'Authorization': 'Bearer $_token',
        'Accept': 'application/vnd.github.v3+json',
        'User-Agent': 'AbsoluteScenes-Mobile',
        'Content-Type': 'application/json',
      },
      body: json.encode(updateData),
    );

    if (response.statusCode != 200 && response.statusCode != 201) {
      final errorData = json.decode(response.body);
      throw Exception(errorData['message'] ?? 'Failed to save to repository: ${response.statusCode}');
    }
  }

  // Clear authentication
  Future<void> clearAuth() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_tokenKey);
    await prefs.remove(_userKey);
    
    _token = null;
    _userInfo = null;
  }
}

class Repository {
  final String fullName;
  final String name;
  final String? description;
  final String bookFileName;

  Repository({
    required this.fullName,
    required this.name,
    this.description,
    required this.bookFileName,
  });
}

File: src/utils/gitHubService.js
/**
 * GitHub Service for AbsoluteScenes Mobile
 * Handles authentication and repository discovery.
 */
import { detectRepoLayout } from '@absolute-scenes/git-sync';

// Runs `mapper` over `items` with at most `limit` in flight at once,
// returning results in the same order as `items` regardless of which
// finishes first. Each `detectRepoLayout` check is itself 3 sequential
// GitHub API calls, and a repo list can have up to 100 entries -- a plain
// sequential loop means up to 300 serialized round trips before the repo
// list screen can render. A bounded pool gets most of the speed of full
// concurrency without firing 100 requests at GitHub simultaneously.
async function mapWithConcurrency(items, limit, mapper) {
  const results = new Array(items.length);
  let nextIndex = 0;

  async function worker() {
    while (nextIndex < items.length) {
      const current = nextIndex++;
      results[current] = await mapper(items[current], current);
    }
  }

  const workerCount = Math.min(limit, items.length);
  await Promise.all(Array.from({ length: workerCount }, worker));
  return results;
}

const REPO_LAYOUT_CHECK_CONCURRENCY = 5;
const REPO_LAYOUT_CACHE_KEY = 'absolute-scenes-mobile:repoLayoutCache';

// GitHub's own repo-list response already includes `pushed_at` for every
// repo -- if it's unchanged since the last time we checked, nothing could
// have changed at the tip of that repo's default branch, so the cached
// layout classification is still correct and detectRepoLayout's 3 API
// calls can be skipped entirely for that repo.
function loadRepoLayoutCache() {
  try {
    return JSON.parse(localStorage.getItem(REPO_LAYOUT_CACHE_KEY)) ?? {};
  } catch (error) {
    console.warn('Failed to read repo layout cache:', error);
    return {};
  }
}

function saveRepoLayoutCache(cache) {
  try {
    localStorage.setItem(REPO_LAYOUT_CACHE_KEY, JSON.stringify(cache));
  } catch (error) {
    console.warn('Failed to persist repo layout cache:', error);
  }
}

class GitHubService {
  constructor() {
    this.token = null;
    this.userInfo = null;
    this.loadStoredAuth();
  }

  loadStoredAuth() {
    try {
      const storedAuth = localStorage.getItem('github_auth');
      if (storedAuth) {
        const authData = JSON.parse(storedAuth);
        this.token = authData.token;
        this.userInfo = authData.userInfo;
        return true;
      }
    } catch (error) {
      console.warn('Failed to load stored GitHub auth:', error);
    }
    return false;
  }

  storeAuth(token, userInfo) {
    try {
      this.token = token;
      this.userInfo = userInfo;
      localStorage.setItem('github_auth', JSON.stringify({ token, userInfo }));
    } catch (error) {
      console.error('Failed to store GitHub auth:', error);
    }
  }

  clearAuth() {
    this.token = null;
    this.userInfo = null;
    localStorage.removeItem('github_auth');
  }

  isAuthenticated() {
    return !!this.token;
  }

  getUserInfo() {
    return this.userInfo;
  }

  async validateAndSetupToken(token) {
    if (!token || !token.startsWith('ghp_')) {
      throw new Error('Please enter a valid GitHub personal access token');
    }

    const response = await fetch('https://api.github.com/user', {
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: 'application/vnd.github.v3+json',
        'User-Agent': 'AbsoluteScenes-Mobile'
      }
    });

    if (response.status === 401) {
      throw new Error('Invalid token. Please check that you copied it correctly.');
    } else if (response.status === 403) {
      throw new Error('Token lacks required permissions. Please ensure "repo" scope is selected.');
    } else if (!response.ok) {
      throw new Error(`GitHub API error: ${response.status}`);
    }

    const userInfo = await response.json();
    this.storeAuth(token, userInfo);
    return userInfo;
  }

  /**
   * Repositories that already hold an AbsoluteScenes book, in either the
   * git-native layout (book.json at root) or the legacy single-.book-file
   * layout that gets migrated automatically on first sync.
   */
  async getUserRepositoriesWithBooks() {
    if (!this.isAuthenticated()) {
      throw new Error('Not authenticated');
    }

    const response = await fetch(
      'https://api.github.com/user/repos?sort=updated&per_page=100',
      {
        headers: {
          Authorization: `Bearer ${this.token}`,
          Accept: 'application/vnd.github.v3+json',
          'User-Agent': 'AbsoluteScenes-Mobile'
        }
      }
    );

    if (!response.ok) {
      throw new Error(`Failed to fetch repositories: ${response.status}`);
    }

    const repos = await response.json();
    const cache = loadRepoLayoutCache();
    const nextCache = {};

    const checked = await mapWithConcurrency(repos, REPO_LAYOUT_CHECK_CONCURRENCY, async repo => {
      const branch = repo.default_branch || 'main';
      const cached = cache[repo.full_name];
      let layout;

      if (cached && cached.pushedAt === repo.pushed_at) {
        layout = cached.layout;
      } else {
        try {
          layout = await detectRepoLayout({
            repo: repo.full_name,
            token: this.token,
            branch
          });
        } catch (error) {
          console.warn(`Skipping ${repo.full_name}: failed to detect layout`, error);
          return null;
        }
      }

      nextCache[repo.full_name] = { layout, pushedAt: repo.pushed_at };

      if (layout !== 'legacy' && layout !== 'new') return null;
      return {
        fullName: repo.full_name,
        name: repo.name,
        description: repo.description,
        defaultBranch: branch
      };
    });

    saveRepoLayoutCache(nextCache);
    return checked.filter(Boolean);
  }
}

export default new GitHubService();
